runtime/shared_heap.c

```diff
@@ -293,24 +294,23 @@ static intnat pool_sweep(struct caml_heap_state* local,
 
 /* Adopt pool from the pool_freelist avail and full pools
    to satisfy an allocation */
-CAMLno_tsan /* Disable TSan reports from this function (see #11040) */
 static pool* pool_global_adopt(struct caml_heap_state* local, sizeclass sz)
 {
   pool* r = NULL;
   int adopted_pool = 0;
 
   /* probably no available pools out there to be had */
-  if( !pool_freelist.global_avail_pools[sz] &&
-      !pool_freelist.global_full_pools[sz] )
+  if( !atomic_load_acquire(&pool_freelist.global_avail_pools[sz]) &&
+      !atomic_load_acquire(&pool_freelist.global_full_pools[sz]) )
```

Comment:
I would expect a `relaxed` here if the intention is to have a fast, approximate check that corresponds to the previous version. `acquire` is probably fine though.

Reply:
My intention was to avoid taking the lock if not necessary, but now that I have to argue about it, I'm no longer sure it was achieving it. Relaxed operation will be fine.